src/core/bitop.d

```diff
@@ -142,6 +156,110 @@ unittest
     assert(test_ctfe == 63);
 }
 
+private alias softBsf(N) = softScan!(N, true);
+private alias softBsr(N) = softScan!(N, false);
+
+/* Shared software fallback implementation for bit scan foward and reverse.
+
+If forward is true, bsf is computed (the index of the first set bit).
+If forward is false, bsr is computed (the index of the last set bit).
+
+-1 is returned if no bits are set (v == 0).
+*/
+private int softScan(N, bool forward)(N v) pure
+    if(is(N == uint) || is(N == ulong))
+{
+    // bsf() and bsr() are officially undefined for v == 0.
+    if (!v)
+        return -1;
+
+    // This is essentially an unrolled binary search:
+    enum mask(ulong lo) = forward? cast(N) lo : cast(N)~lo;
+    enum inc(int up) = forward? up : -up;
```

Comment:
> What is this, ruby? :-)

For what it's worth, I've never written so much as one line of ruby in my life. I do like the ternary operator, though.